modules/storage/src/service/fs.rs

```diff
@@ -0,0 +1,225 @@
+
+/// A filesystem backed store
+///
+/// ## Layout
+///
+/// The layout of the filesystem is as follows:
+///
+/// ```ignore
+/// <base>/
+///   content/
+///     <digest[0..2]>/
+///       <digest[2..4]>/
+///         <digest> # file
+/// ```
+///
+/// The idea behind that is to limit the number of directory entries. For some filesystems,
+/// the performance can degrade is directories get too big (have too many entries).
+///
+/// This layout limits the number of entries on the first two layers to 256, and limits the chance
+/// of a file ending up in the same directory by 65536. Assuming an average distribution of hashes,
+/// that would allow us to store 16777216 files, until we hit the cap of 256 entries in the lowest
+/// lever directories. Should we need to increase that limit, we could easily add an additional
+/// layer.
+#[derive(Clone, Debug)]
+pub struct FileSystemBackend {
+    content: PathBuf,
+}
+
+const NUM_LEVELS: usize = 2;
+
+impl FileSystemBackend {
+    pub async fn new(base: impl Into<PathBuf>) -> anyhow::Result<Self> {
+        let base = base.into();
+        let content = base.join("content");
+
+        create_dir(&content)
+            .await
+            .or_else(|err| {
+                if err.kind() == ErrorKind::AlreadyExists {
+                    Ok(())
+                } else {
+                    Err(err)
+                }
+            })
+            .with_context(|| {
+                format!(
+                    "unable to create 'content' directory in the file system base: {}",
+                    base.display()
+                )
+            })?;
+
+        Ok(Self { content })
+    }
+
+    /// Create a new storage for testing
+    pub async fn for_test() -> anyhow::Result<(Self, TempDir)> {
+        let dir = tempdir()?;
+
+        Self::new(dir.path()).await.map(|result| (result, dir))
+    }
+}
+
+impl StorageBackend for FileSystemBackend {
+    type Error = std::io::Error;
+
+    async fn store<E, S>(&self, stream: S) -> Result<Output<Sha256>, StoreError<E, Self::Error>>
+    where
+        E: Debug,
+        S: Stream<Item = Result<Bytes, E>>,
+    {
+        // create a new temp file
+
+        let mut file = File::from(tempfile().map_err(StoreError::Backend)?);
+
+        // set up reader
+
+        let mut stream = pin!(stream);
+        let mut digest = Sha256::new();
+
+        // process reader
+
+        while let Some(next) = stream
+            .next()
+            .await
+            .transpose()
+            .map_err(StoreError::Stream)?
+        {
+            digest.update(&next);
+            file.write_all(&next).await.map_err(StoreError::Backend)?;
+        }
+
+        // finalize the digest
+
+        let digest = digest.finalize();
+
+        // create the target path
+
+        let hash = hex::encode(digest);
+        let target = level_dir(&self.content, &hash, NUM_LEVELS);
+        create_dir_all(&target).await.map_err(StoreError::Backend)?;
```

Comment:
Should we sprinkle some `from` derives or impl `From<..>` to avoid having to `map_err(...)`?

Reply:
Probably. I expect much of the storage module to evolve.